README.md
## Algebird

[![Build status](https://img.shields.io/travis/twitter/algebird/develop.svg)](http://travis-ci.org/twitter/algebird)
[![Codecov branch](https://img.shields.io/codecov/c/github/twitter/algebird/develop.svg?maxAge=3600)](https://codecov.io/github/twitter/algebird)
[![Latest version](https://index.scala-lang.org/twitter/algebird/algebird-core/latest.svg?color=orange)](https://index.scala-lang.org/twitter/algebird/algebird-core)
[![Chat](https://badges.gitter.im/twitter/algebird.svg)](https://gitter.im/twitter/algebird?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

### Overview

Abstract algebra for Scala. This code is targeted at building aggregation systems (via [Scalding](https://github.com/twitter/scalding) or [Apache Storm](http://storm.apache.org/)). It was originally developed as part of Scalding's Matrix API, where Matrices had values which are elements of Monoids, Groups, or Rings. Subsequently, it was clear that the code had broader application within Scalding and on other projects within Twitter.

See the [Algebird website](https://twitter.github.io/algebird) for more information.

### What can you do with this code?

```scala
> ./sbt algebird-core/console

Welcome to Scala version 2.10.5 (Java HotSpot(TM) 64-Bit Server VM, Java 1.7.0_40).
Type in expressions to have them evaluated.
Type :help for more information.

scala> import com.twitter.algebird._
import com.twitter.algebird._

scala> import com.twitter.algebird.Operators._
import com.twitter.algebird.Operators._

scala> Map(1 -> Max(2)) + Map(1 -> Max(3)) + Map(2 -> Max(4))
res0: scala.collection.immutable.Map[Int,com.twitter.algebird.Max[Int]] = Map(2 -> Max(4), 1 -> Max(3))
```

In the above, the class `Max[T]` signifies that the `+` operator should actually be `max` (this is accomplished by providing an implicit instance of a typeclass for `Max` that handles `+`).

- Model a wide class of "reductions" as a sum on some iterator of a particular value type. For example, average, moving average, max/min, set union, approximate set size (in much less memory with HyperLogLog), approximate item counting (using CountMinSketch).
- All of these combine naturally in tuples, vectors, maps, options and more standard scala classes.
- Implementations of Monoids for interesting approximation algorithms, such as Bloom filter, HyperLogLog and CountMinSketch. These allow you to think of these sophisticated operations like you might numbers, and add them up in hadoop or online to produce powerful statistics and analytics.

## Documentation

To learn more and find links to tutorials and information around the web, check out the [Algebird website](https://twitter.github.io/algebird).

The latest API docs are hosted on Algebird's [ScalaDoc index](https://twitter.github.io/algebird/api/).

## Get Involved + Code of Conduct

Pull requests and bug reports are always welcome! Check out our [Contributing guide](https://twitter.github.io/algebird/contributing.html) for information on what we most need help with and how you can get started contributing.

Discussion occurs primarily on the Algebird Gitter channel: [![Chat](https://badges.gitter.im/twitter/algebird.svg)](https://gitter.im/twitter/algebird?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

We also monitor the [Algebird mailing list](https://groups.google.com/forum/#!forum/algebird).

Issues should be reported on the [GitHub issue tracker](https://github.com/twitter/algebird/issues).

We use a lightweight form of project governance inspired by the one used by Apache projects.

Please see [Contributing and Committership](https://github.com/twitter/analytics-infra-governance#contributing-and-committership) for our code of conduct and our pull request review process.

The TL;DR is send us a pull request, iterate on the feedback + discussion, and get a +1 from a [Committer](COMMITTERS.md) in order to get your PR accepted.

The current list of active committers (who can +1 a pull request) can be found here: [Committers](COMMITTERS.md)

A list of contributors to the project can be found here: [Contributors](https://github.com/twitter/algebird/graphs/contributors)

## Maven

Algebird modules are available on maven central. The current groupid and version for all modules is, respectively, `"com.twitter"` and  `0.12.3`.

See [Algebird's page on the Scaladex](https://index.scala-lang.org/twitter/algebird) for information on all published artifacts and their associated Scala versions. Algebird currently supports Scala 2.10 and 2.11.

## Projects using Algebird

- [Scalding](http://github.com/twitter/scalding)
- [Spark](https://github.com/mesos/spark/pull/480)
- [Simmer](https://github.com/avibryant/simmer)
- [Summingbird](https://github.com/twitter/summingbird)
- [Packetloop](https://www.packetloop.com) (see [this tweet](https://twitter.com/cloudjunky/status/355073917720858626)
- Ebay uses Algebird for machine learning: [ScalaDays talk](http://www.slideshare.net/VitalyGordon/scalable-and-flexible-machine-learning-with-scala-linkedin)

Other projects built with Algebird, as compiled by the Scaladex: [![Scaladex Dependents](https://index.scala-lang.org/count.svg?q=dependencies:twitter/algebird*&subject=scaladex:&color=blue&style=flat-square)](https://index.scala-lang.org/search?q=dependencies:twitter/algebird-core)

## Authors

* Oscar Boykin <http://twitter.com/posco>
* Avi Bryant <http://twitter.com/avibryant>
* Edwin Chen <http://twitter.com/echen>
* ellchow <http://github.com/ellchow>
* Mike Gagnon <https://twitter.com/gmike>
* Moses Nakamura <https://twitter.com/mnnakamura>
* Steven Noble <http://twitter.com/snoble>
* Sam Ritchie <http://twitter.com/sritchie>
* Ashutosh Singhal <http://twitter.com/daashu>
* Argyris Zymnis <http://twitter.com/argyris>

## License

Copyright 2016 Twitter, Inc.

Licensed under the [Apache License, Version 2.0](http://www.apache.org/licenses/LICENSE-2.0).


CHANGES.md
# Algebird #

### Version 0.12.3 (Not yet released) ###

* Add `SetDiff` data structure to `algebird-core`: https://github.com/twitter/algebird/pull/555
* Add `Ring[BigDecimal]`, modeled after `Ring[BigInt]`: https://github.com/twitter/algebird/pull/553
* "Exponential Histogram" sliding window counter implementation added to `algebird-core` as `ExpHist`: https://github.com/twitter/algebird/pull/568
* improve HLLSeries performance: https://github.com/twitter/algebird/pull/575
* Add a microsite at <https://twitter.github.io/algebird> via the `sbt-microsites` plugin, along with docs for all typeclasses and data structures: https://github.com/twitter/algebird/pull/576
* Adds lots of scalacheck `Arbitrary` and `Gen` instances to `algebird-test`, under `com.twitter.algebird.scalacheck.{ gen, arbitrary }`: https://github.com/twitter/algebird/pull/579
* Add `Monoid[Max[Vector[T]]]`, `Monoid[Max[Stream[T]]]`: https://github.com/twitter/algebird/pull/579
* Add `FirstAggregator` and `LastAggregator`, and docs and API / perf improvements for `First`, `Last`, `Min`, `Max`: https://github.com/twitter/algebird/pull/579
* Add `LawsEquiv` versions of all laws: https://github.com/twitter/algebird/pull/584
* Deprecates broken group/ring for `Future`/`Try`: https://github.com/twitter/algebird/pull/584
* Add `metricsLaws[T]` to `BaseProperties` in `algebird-test`: https://github.com/twitter/algebird/pull/584
* Modify generated `Tuple2Monoid`, etc to extend `TupleNSemigroup`, giving subclasses access to efficient `sumOption`: https://github.com/twitter/algebird/pull/585
* optimize `Generated{Abstract,Product}Algebra.sumOption` with benchmarking https://github.com/twitter/algebird/pull/591
* Add an efficient `sumOption`, `+`, `-`, methods and docs to `AveragedValue`: https://github.com/twitter/algebird/pull/589

### Version 0.12.2 ###

* optimize `CMS.create(Seq[K])`: https://github.com/twitter/algebird/pull/537
* Update to sbt 0.13.11: https://github.com/twitter/algebird/pull/540
* Add `Semigroup.maybePlus`: https://github.com/twitter/algebird/pull/544
* Improve the build and test times and add Codecov.io: https://github.com/twitter/algebird/pull/541

### Version 0.12.1 ###

* Add `Identity` class and `Monad` instance: https://github.com/twitter/algebird/pull/511
* Using `dropRight` instead of `tail` for shortening `BytesSpec` array: https://github.com/twitter/algebird/pull/510
* Fix flaky `MinHasherSpec` test: https://github.com/twitter/algebird/pull/514
* fix input type of `toRichTraversable`: https://github.com/twitter/algebird/pull/518
* build in scala 2.11 by default: https://github.com/twitter/algebird/pull/520
* make `SSOne`/`SSMany` constructors private and add apply method with count: https://github.com/twitter/algebird/pull/519
* Fix `BytesSpec` to not reference head on an empty list  https://github.com/twitter/algebird/pull/525
* Add an `Aggregator.randomSample` aggregator: https://github.com/twitter/algebird/pull/529
* Add `sortedTakeBy` and `sortedReverseTakeBy` to `Aggregator.scala`: https://github.com/twitter/algebird/pull/527
* Add `Batched[A]` type for efficient lazy addition: https://github.com/twitter/algebird/pull/530
* Add a default `k` value for `Aggregator.approximatePercentile`: https://github.com/twitter/algebird/pull/531

### Version 0.12.0 ###

* Implement an appendMonoid Aggregator factory which yields aggregators…: https://github.com/twitter/algebird/pull/501
* Dealing with probabilistic tests: https://github.com/twitter/algebird/pull/478
* Add Applicative.sequenceGen: https://github.com/twitter/algebird/pull/498
* Create a sparse Count-Min-Sketch.: https://github.com/twitter/algebird/pull/464
* fix name and visibility of negativePowersOfTwo: https://github.com/twitter/algebird/pull/492
* Speed up HLL presentation by 100x: https://github.com/twitter/algebird/pull/491
* Test Semigroup#sumOption using Iterator instead of List: https://github.com/twitter/algebird/pull/490
* Fix tests that were not actually running: https://github.com/twitter/algebird/pull/485
* add immutable version of sorted(Reverse)Take: https://github.com/twitter/algebird/pull/484
* Cuber/roller macros: https://github.com/twitter/algebird/pull/483
* Add sanity requirement for Approximate: https://github.com/twitter/algebird/pull/481
* Ioconnell/make develop version have snapshot suffix: https://github.com/twitter/algebird/pull/482
* Upgrade scalacheck and scalatest: https://github.com/twitter/algebird/pull/480
* Adding scoped top-N CMS monoid: https://github.com/twitter/algebird/pull/471
* Fix Qtree quantileBounds off-by-one error: https://github.com/twitter/algebird/pull/472
* Move benchmarks to JMH: https://github.com/twitter/algebird/pull/473
* Ianoc/q tree benchmark more coverage: https://github.com/twitter/algebird/pull/474
* Optimize QTree a bunch: https://github.com/twitter/algebird/pull/475
* Disable coveralls, shows up as builds pending that are long finished: https://github.com/twitter/algebird/pull/476

### Version 0.11.0 ###
* Move CMSHasherByteArray from scalding: https://github.com/twitter/algebird/pull/467
* Upgrade sbt launcher script (sbt-extras): https://github.com/twitter/algebird/pull/469
* Create case class macros for algebraic structures: https://github.com/twitter/algebird/pull/466
* Refactor MapAggregator: https://github.com/twitter/algebird/pull/462
* Algebird support for spark: https://github.com/twitter/algebird/pull/397
* Add MapAggregator from 1 (key, aggregator) pair: https://github.com/twitter/algebird/pull/452
* Remove unnecessary use of scala.math: https://github.com/twitter/algebird/pull/455
* Don&#39;t call deprecated HyperLogLog methods in tests: https://github.com/twitter/algebird/pull/456
* Update product_generators.rb: https://github.com/twitter/algebird/pull/457
* Pzheng/gaussian euclidean: https://github.com/twitter/algebird/pull/448

### Version 0.10.2 ###
* QTree quantileBounds assert percentile <= 1.0 #447

### Version 0.10.1 ###

project/Build.scala
package algebird

import sbt._
import Keys._
import com.typesafe.tools.mima.plugin.MimaPlugin.mimaDefaultSettings
import com.typesafe.tools.mima.plugin.MimaKeys.previousArtifact
import scalariform.formatter.preferences._
import com.typesafe.sbt.SbtScalariform._
import pl.project13.scala.sbt.JmhPlugin

object AlgebirdBuild extends Build {

  val paradiseVersion = "2.0.1"
  val quasiquotesVersion = "2.0.1"

  def scalaBinaryVersion(scalaVersion: String) = scalaVersion match {
    case version if version startsWith "2.10" => "2.10"
    case version if version startsWith "2.11" => "2.11"
    case _ => sys.error("unknown error")
  }
  def isScala210x(scalaVersion: String) = scalaBinaryVersion(scalaVersion) == "2.10"

  val sharedSettings = Project.defaultSettings ++ scalariformSettings ++  Seq(
    organization := "com.twitter",
    scalaVersion := "2.10.5",
    crossScalaVersions := Seq("2.10.5", "2.11.7"),
    ScalariformKeys.preferences := formattingPreferences,

    resolvers ++= Seq(
      "snapshots" at "http://oss.sonatype.org/content/repositories/snapshots",
      "releases"  at "http://oss.sonatype.org/content/repositories/releases"
    ),

    parallelExecution in Test := true,

    javacOptions ++= Seq("-target", "1.6", "-source", "1.6"),

    scalacOptions ++= Seq("-unchecked", "-deprecation", "-optimize", "-Xlint", "-language:implicitConversions", "-language:higherKinds", "-language:existentials"),

    scalacOptions <++= (scalaVersion) map { sv =>
        if (sv startsWith "2.10")
          Seq("-Xdivergence211")
        else
          Seq()
    },

    javacOptions ++= Seq("-target", "1.6", "-source", "1.6"),

    libraryDependencies += "junit" % "junit" % "4.11" % "test",

    // Publishing options:
    publishMavenStyle := true,

    publishArtifact in Test := false,

    pomIncludeRepository := { x => false },

    publishTo <<= version { v =>
      Some(
        if (v.trim.endsWith("SNAPSHOT"))
          Opts.resolver.sonatypeSnapshots
        else
          Opts.resolver.sonatypeStaging
      )
    },

    pomExtra := (
      <url>https://github.com/twitter/algebird</url>
      <licenses>
        <license>
          <name>Apache 2</name>
          <url>http://www.apache.org/licenses/LICENSE-2.0.txt</url>
          <distribution>repo</distribution>
          <comments>A business-friendly OSS license</comments>
        </license>
      </licenses>
      <scm>
        <url>git@github.com:twitter/algebird.git</url>
        <connection>scm:git:git@github.com:twitter/algebird.git</connection>
      </scm>
      <developers>
        <developer>
          <id>oscar</id>
          <name>Oscar Boykin</name>
          <url>http://twitter.com/posco</url>
        </developer>
        <developer>
          <id>sritchie</id>
          <name>Sam Ritchie</name>
          <url>http://twitter.com/sritchie</url>
        </developer>
      </developers>)
  ) ++ mimaDefaultSettings


  lazy val formattingPreferences = {
    import scalariform.formatter.preferences._
    FormattingPreferences().
      setPreference(AlignParameters, false).
      setPreference(PreserveSpaceBeforeArguments, true)
  }

  /**
    * This returns the youngest jar we released that is compatible with
    * the current.
    */
  val unreleasedModules = Set[String]()

  def youngestForwardCompatible(subProj: String) =
    Some(subProj)
      .filterNot(unreleasedModules.contains(_))
      .map { s => "com.twitter" % ("algebird-" + s + "_2.10") % "0.11.0" }

  lazy val algebird = Project(
    id = "algebird",
    base = file("."),
    settings = sharedSettings ++ DocGen.publishSettings
    ).settings(
    test := { },
    publish := { }, // skip publishing for this root project.
    publishLocal := { }
  ).aggregate(
    algebirdTest,
    algebirdCore,
    algebirdUtil,
    algebirdBijection,
    algebirdBenchmark,
    algebirdSpark
  )

  def module(name: String) = {
    val id = "algebird-%s".format(name)
    Project(id = id, base = file(id), settings = sharedSettings ++ Seq(
      Keys.name := id,
      previousArtifact := youngestForwardCompatible(name))
    )
  }

  lazy val algebirdCore = module("core").settings(
    test := { }, // All tests reside in algebirdTest
    initialCommands := """
                       import com.twitter.algebird._
                       """.stripMargin('|'),
    libraryDependencies <++= (scalaVersion) { scalaVersion =>
      Seq("com.googlecode.javaewah" % "JavaEWAH" % "0.6.6",
          "org.scala-lang" % "scala-reflect" % scalaVersion) ++ {
        if (isScala210x(scalaVersion))
          Seq("org.scalamacros" %% "quasiquotes" % quasiquotesVersion)
        else
          Seq()
      }
    },
    libraryDependencies += "org.apache.commons" % "commons-math3" % "3.5",
    sourceGenerators in Compile <+= sourceManaged in Compile map { outDir: File =>
      GenTupleAggregators.gen(outDir)
    }, addCompilerPlugin("org.scalamacros" % "paradise" % paradiseVersion cross CrossVersion.full)
  )

  lazy val algebirdTest = module("test").settings(
    testOptions in Test ++= Seq(Tests.Argument(TestFrameworks.ScalaCheck, "-verbosity", "4")),
    libraryDependencies <++= (scalaVersion) { scalaVersion =>
      Seq("org.scalacheck" %% "scalacheck" % "1.12.5",
          "org.scalatest" %% "scalatest" % "2.2.4") ++ {
        if (isScala210x(scalaVersion))
          Seq("org.scalamacros" %% "quasiquotes" % quasiquotesVersion)
        else
          Seq()
      }
    },
    libraryDependencies += "org.apache.commons" % "commons-math3" % "3.5",
    addCompilerPlugin("org.scalamacros" % "paradise" % paradiseVersion cross CrossVersion.full)
  ).dependsOn(algebirdCore)

  lazy val algebirdBenchmark = module("benchmark").settings(JmhPlugin.projectSettings:_*).settings(
     libraryDependencies ++= Seq(
       "com.twitter" %% "bijection-core" % "0.8.0",
       "org.apache.commons" % "commons-math3" % "3.5")
  ).dependsOn(algebirdCore, algebirdUtil, algebirdTest % "test->compile").enablePlugins(JmhPlugin)

  lazy val algebirdUtil = module("util").settings(
    libraryDependencies += "com.twitter" %% "util-core" % "6.20.0"
  ).dependsOn(algebirdCore, algebirdTest % "test->test")

  lazy val algebirdBijection = module("bijection").settings(
    libraryDependencies += "com.twitter" %% "bijection-core" % "0.8.0"
  ).dependsOn(algebirdCore, algebirdTest % "test->test")

  lazy val algebirdSpark = module("spark").settings(
    libraryDependencies += "org.apache.spark" %% "spark-core" % "1.3.0" % "provided"
  ).dependsOn(algebirdCore, algebirdTest % "test->test")
}


